File: scratch/get_scopes.py
import urllib.request
import urllib.parse
import json

# Get admin token
url = "http://localhost:8081/realms/master/protocol/openid-connect/token"
data = urllib.parse.urlencode({
    "client_id": "admin-cli",
    "username": "admin",
    "password": "admin",
    "grant_type": "password"
}).encode("utf-8")

req = urllib.request.Request(url, data=data)
try:
    with urllib.request.urlopen(req) as response:
        res = json.loads(response.read().decode())
        token = res["access_token"]
except Exception as e:
    print("Failed to get token:", e)
    exit(1)

# Get client scopes for josanz-web-app-realm
scopes_url = "http://localhost:8081/admin/realms/josanz-web-app-realm/client-scopes"
req = urllib.request.Request(scopes_url)
req.add_header("Authorization", f"Bearer {token}")

try:
    with urllib.request.urlopen(req) as response:
        scopes = json.loads(response.read().decode())
        print("Available Client Scopes in Realm:")
        for scope in scopes:
            print(f"- {scope['name']} ({scope['id']})")
except Exception as e:
    print("Failed to get scopes:", e)

File: scratch/get_scopes.js
const http = require('http');
const querystring = require('querystring');

function post(url, data) {
  return new Promise((resolve, reject) => {
    const urlObj = new URL(url);
    const postData = querystring.stringify(data);
    const options = {
      hostname: urlObj.hostname,
      port: urlObj.port,
      path: urlObj.pathname,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': Buffer.byteLength(postData)
      }
    };

    const req = http.request(options, (res) => {
      let body = '';
      res.on('data', (chunk) => body += chunk);
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve(JSON.parse(body));
        } else {
          reject(new Error(`Status ${res.statusCode}: ${body}`));
        }
      });
    });

    req.on('error', reject);
    req.write(postData);
    req.end();
  });
}

function decodeToken(token) {
  const parts = token.split('.');
  if (parts.length < 2) return null;
  return JSON.parse(Buffer.from(parts[1], 'base64').toString());
}

async function main() {
  try {
    const tokenRes = await post('http://localhost:8081/realms/josanz-web-app-realm/protocol/openid-connect/token', {
      client_id: 'josanz-web-app-spa',
      username: 'admin',
      password: 'admin',
      grant_type: 'password',
      scope: 'openid email profile'
    });
    console.log('Token Response successfully obtained.');
    console.log('\nDecoded Access Token Payload:');
    console.log(JSON.stringify(decodeToken(tokenRes.access_token), null, 2));
  } catch (err) {
    console.error(err);
  }
}

main();
